packages/apps/plugins/plugin-space/src/components/SpacePresence.tsx

```diff
@@ -2,118 +2,176 @@
-      data: { spaceKey: space!.key.toHex() },
-    });
-  };
+  const members = spaceMembers
+    .filter((member) => member.presence === 1 && !identity.identityKey.equals(member.identity.identityKey))
+    .map((member) => ({
+      ...member,
+      match: viewers.has(member.identity.identityKey),
+      lastSeen: viewers.get(member.identity.identityKey) ?? Infinity,
+    }))
+    .toSorted((a, b) => a.lastSeen - b.lastSeen);
 
-  if (!space || defaultSpace?.key.equals(space.key)) {
-    return null;
-  }
-
-  const viewers = spacePlugin.provides.space.viewers.filter((viewer) => {
-    return space.key.equals(viewer.spaceKey) && object.id === viewer.objectId && Date.now() - viewer.lastSeen < 30_000;
-  });
+  const onMoreClick = () => client.shell.shareSpace({ spaceKey: space.key });
 
-  return (
-    <ObjectPresence
-      size={density === 'fine' ? 2 : 4}
-      classNames={density === 'fine' ? 'is-6' : 'pli-2.5'}
-      onShareClick={handleShare}
-      viewers={viewers}
-    />
+  return density === 'fine' ? (
+    <SmallPresence members={members.filter((member) => member.match)} />
+  ) : (
+    <FullPresence members={members} onMoreClick={onMoreClick} />
   );
 };
 
-export type ObjectPresenceProps = ThemedClassName<{
+export type Member = SpaceMember & {
+  /**
+   * True if the member is currently viewing the specified object.
+   */
+  match: boolean;
+
+  /**
+   * Last time a member was seen on this object.
+   */
+  lastSeen: number;
+};
+
+export type MemberPresenceProps = ThemedClassName<{
   size?: Size;
-  viewers?: ObjectViewer[];
+  members?: Member[];
   showCount?: boolean;
-  onShareClick?: () => void;
+  onMemberClick?: (member: Member) => void;
+  onMoreClick?: () => void;
 }>;
 
-export const ObjectPresence = (props: ObjectPresenceProps) => {
-  const {
-    onShareClick,
-    viewers = [],
-    size = 4,
-    showCount = !props?.size || (props.size !== 'px' && props.size > 4),
-    classNames,
-  } = props;
+export const FullPresence = (props: MemberPresenceProps) => {
+  const { members = [], size = 9, onMemberClick, onMoreClick } = props;
   const { t } = useTranslation(SPACE_PLUGIN);
-  const density = useDensityContext();
-  return density === 'fine' && viewers.length === 0 ? (
+
+  if (members.length === 0) {
+    return null;
+  }
+
+  return (
+    <AvatarGroup.Root size={size} classNames='mbs-2 mie-4'>
+      {members.slice(0, 3).map((member, i) => {
+        const memberHex = member.identity.identityKey.toHex();
+        const status = member.match ? 'current' : 'active';
+        const name = member.identity.profile?.displayName ?? generateName(member.identity.identityKey.toHex());
+        return (
+          <Tooltip.Root key={memberHex}>
+            <Tooltip.Trigger>
+              <AvatarGroupItem.Root color={getColorForValue({ value: memberHex, type: 'color' })} status={status}>
+                <Avatar.Frame style={{ zIndex: members.length - i }} onClick={() => onMemberClick?.(member)}>
```

Comment:
There’s a way to handle this without messing with `zIndex`, but I forget… if this didn’t work out of the box without doing this, this is fine for now, it just may cause problems because there are other elements on the page with low z-indexes.

Reply:
This was how this control was done previously, I didn't add this so I'm not exactly sure of it's intended purpose.